MemberClusterAnnounce webhook and Clusterset status implementations

aravindakidambi opened the pull request:
MCS: MemberClusterAnnounce webhook and CluterSet.Status implementation   
   
MemberClusterAnnounce webhook
-----------------------------------
Implement a custom webhook to authenticate the request
to write MemberClusterAnnouce.
The webhook will get ClusterID from MemberClusterAnnounce
and get the ServiceAccount associated with that member cluster
from ClusterSet configuration and ensure the incoming request
has been authenticated with the same ServiceAccount.
This prevents a member spoofing the ClusterID when writing
MemberClusterAnnounce.
    
NOTE: Removed the default webhook from kube-builder
because it does not provide access to the user-info
of the request.

ClusterSet.Status computation on leader clusters
--------------------------------------------------
MemberClusterAnnounceReconciler will run an interval timer
which monitors MemberClusterAnnounce resources to compute
status as follows:
1. All configured members will start with:
   a. "Ready" = "Unknown"
   b. "ImportsResources" = "False"
2. The timer determines the following:
   a. If there is atleast one update of MemberClusterAnnounce
      resource from a member in a consecutive 3 timer intervals,
      the status of member will be "Ready" = "True"
   b. If "Ready" = "True", based on the last MemberClusterAnnounce
      received from that member:
      i. "ImportsResources" = "True" if the local leader cluster
         is the elected leader for member. Message will be
         "Local cluster is the elected leader of member <cluster-id>"
         and reason says "ElectedLeader"
      ii. "ImportsResources" = "False" if some other leader
          was elected by this member. Message will be
          "Local cluster is not the elected leader of member <cluster-id>"
          and reason will be "NotElectedLeader"
      iii. "ImportsResources" = "Unknown", if no leader has
           been elected yet by this member. Message will be
           "Leader has not been elected yet"
   c. If there is no update in a consecutive 3 timer interval, then the
      member is marked "Ready" = "False" and "ImportsResources" = "False".
      Message will state - "No MemberClusterAnnounce update after <timestamp>" and
      Reason will be "Disconnected".

LeaderClusterSetReconciler will have a goroutine that runs
periodically to update the ClusterSet.Status, as follows:
1. TotalClusters is the number of clusters in the ClusterSet resource
   last processed.
2. ObservedGeneration is the Generation in the ClusterSet resource
   last processed.
3. Individual cluster status is obtained from MemberClusterAnnounceReconciler.
4. ReadyClusters is the number of clusters with "Ready" = "True".
5. Overall condition of the ClusterSet is also computed as follows:
   a. "Ready" = "True" if all clusters have
      "Ready" = "True". Message & Reason will be absent.
   b. "Ready" = "Unknown" if all clusters have
      "Ready" = "Unknown". Message will be "All clusters have an unknown status"
      and Reason will be "NoReadyCluster".
   c. "Ready" = "False" for any other combination of cluster
      statues across all clusters. Message will be empty and Reason
      will be "NoReadyCluster".

Note: Other leaders and local leader cluster will not be present in the status.
      Because its status is not meaningful and can skew the overall status.

ClusterSet.Status computation on member clusters
--------------------------------------------------    
RemoteCommonAreaManager will support getting status of
all leaders from individal RemoteCommonArea.

RemoteCommonArea provides status as follows:
1. When configured, leader start with:
   a. "Ready" = "Unknown"
   b. "IsElectedLeader" = "False"
2. Update "Ready" = "True"/"False" based on connectivity to
   RemoteCommonArea. Message will contain error if member
   has any error with connectivity. Reason will be
   "Disconnected" in error cases.
3. There will be an additional status for "IsElectedLeader"
   and this will be "True" if leader election has finished
   and has elected this leader, "False" if some other leader
   was elected, "Unknown" if leader election is incomplete.
   a. Message will be "This leader cluster is/is not an elected
      leader for local cluster".
   b. If leader election is incomplete,
      Reason = "LeaderElectionInProgress".

MemberClusterSetReconciler will have a goroutine
that runs periodically to update the status, as follows:
1. TotalClusters is updated based on number of clusters in
   the last processed config.
2. ObservedGeneration is from Generation of the last processed
   config.
3. Individual cluster status is obtained from RemoteCommonAreaManager.
4. ReadyClusters is computed based on number of cluster
   with "Ready" = "True".
5. Overall condition of the ClusterSet is computed as follows:
   a. "Ready" = "True" if it is connected to
      atleast one leader and elected it as a leader.
   b. "Ready" = "False" if it is disconnected
      from all leaders.
   c. "Ready" = "Unknown" otherwise. Message
      will be "Leader not elected yet".

luolanzone commented on multicluster/controllers/multicluster/leader_clusterset_controller.go:
```diff
@@ -156,3 +209,68 @@ func validateLocalClusterClaim(c client.Client, clusterSet *multiclusterv1alpha1
+	}
+	if readyClusters == len(r.clusterSetConfig.Spec.Members) {
+		overallCondition.Status = v1.ConditionTrue
+		overallCondition.Reason = ""
```
maybe add a reason like `ClusterSet is healthy` to indicate all clusters are healthy?

luolanzone commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,264 @@ package multicluster
 
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=memberclusterannounces,verbs=get;list;watch;create;update;patch;delete
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=memberclusterannounces/status,verbs=get;update;patch
 //+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=memberclusterannounces/finalizers,verbs=update
```
better to add all necessary marks to run this standalone controller.
I see you will list ClusterClaim in this reconciler. so better  to add below line:
```
//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterclaims,verbs=get;list;watch;

```

aravindakidambi commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,264 @@ package multicluster
+			continue
+		}
+		status := r.memberStatus[member]
+		// Check if the member has connected atleast once in the last 3 intervals.
```
atleast is the correct word I think?

aravindakidambi commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,264 @@ package multicluster
+		// Check if the member has connected atleast once in the last 3 intervals.
+		duration := time.Since(data.lastUpdateTime)
+		if duration <= TimerDuration {
+			// Member has updated MemberClusterStatus atleast once in the last 3 intervals.
```
same here, I think atleast is correct

aravindakidambi commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,264 @@ package multicluster
+	defer r.mapLock.Unlock()
+	r.mapLock.Lock()
+
+	if _, ok := r.memberStatus[MemberId]; ok {
```
Yes, thanks for pointing out.

jianjuns reviewed (commented):
In commit message: inorder -> in order

jianjuns commented on multicluster/cmd/multicluster-controller/memberclusterannounce_webhook.go:
```diff
@@ -0,0 +1,87 @@
+
+	clusterSet := clusterSetList.Items[0]
+	for _, member := range clusterSet.Spec.Members {
+		if clusterSet.Name == memberClusterAnnounce.ClusterSetID && member.ClusterID == memberClusterAnnounce.ClusterID {
```
Simpler to check clusterSet.Name == memberClusterAnnounce.ClusterSetID before the loop?

jianjuns commented on multicluster/cmd/multicluster-controller/memberclusterannounce_webhook.go:
```diff
@@ -0,0 +1,87 @@
+	for _, member := range clusterSet.Spec.Members {
+		if clusterSet.Name == memberClusterAnnounce.ClusterSetID && member.ClusterID == memberClusterAnnounce.ClusterID {
+			// validate the ServiceAccount used is correct
+			if member.ServiceAccount == saName {
```
Question - could not we go this way instead: MemberClusterAnnounce name must be ClusterID, and then when creating ServiceAccount we can limit the ServiceAccount can only create/update MemberClusterAnnounce of name ClusterID?

But I forget if K8s RBAC can restrict Create to a specific name.

aravindakidambi replied:
I looked and couldnt find a way to define Service Account RBAC based on name of a resource. Ill explore it more and update the webhook logic if there is a way. But for now, could we keep this?

jianjuns replied:
For verbs of get/update, we should be able to specify resourceNames in Role rules. But I roughly remember that does not work for create.

I am fine to merge the current way first, but could you verify if resourceNames work with create or not?

aravindakidambi replied:
I see, ok, I see resourceName can be provided, I can test whether it works for create or not.

aravindakidambi replied:
https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources

yes, it can only be used for list/watch, not create/delete.

jianjuns commented on multicluster/apis/multicluster/v1alpha1/clusterset_types.go:
```diff
@@ -73,8 +73,13 @@ type ClusterConditionType string
+	ClusterIsElectedLeader ClusterConditionType = "IsElectedLeader"
+	// ClusterWatchesLocalResources indicates whether the member cluster has elected the
+	// local cluster as the elected leader and watches resources from it.
+	// Used on in Leader clusters.
```
Remove "in".

aravindakidambi replied:
Sure, I fixed it, maybe in reads better than on

jianjuns commented on multicluster/apis/multicluster/v1alpha1/clusterset_types.go:
```diff
@@ -73,8 +73,13 @@ type ClusterConditionType string
+	// ClusterWatchesLocalResources indicates whether the member cluster has elected the
+	// local cluster as the elected leader and watches resources from it.
+	// Used on in Leader clusters.
+	ClusterWatchesLocalResources ClusterConditionType = "WatchesLocalResources"
```
How about "ImportsResources"?

aravindakidambi replied:
Sure

jianjuns commented on multicluster/controllers/multicluster/leader_clusterset_controller.go:
```diff
@@ -68,31 +85,157 @@ func (r *LeaderClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
+// 3. Individual cluster status is obtained from MemberClusterAnnounce
+//    controller
+// 3. ReadyClusters is the number of clusters with "Ready" = "True"
+// 4. Overall condition of the cluster is also computed as follows:
```
cluster -> ClusterSet

aravindakidambi replied:
Done

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,275 @@ package multicluster
 		Complete(r)
 }
+
+func (r *MemberClusterAnnounceReconciler) processMcsStatus() {
```
Probably: Mcs -> MCS

aravindakidambi replied:
Done

jianjuns reviewed (commented):
A few minor comments for the ClusterSet status commit.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -18,45 +18,275 @@ package multicluster
+	}
+}
+
+func (r *MemberClusterAnnounceReconciler) GetMemberClusterStatues() []multiclusterv1alpha1.ClusterStatus {
```
GetMemberClusterStatues -> GetMemberClusterStatuses

aravindakidambi replied:
oops, thanks for catching it. Fixing it.

jianjuns reviewed (commented):
One more.

jianjuns commented:
LGTM. We need approval from @luolanzone

luolanzone commented:
Hi @aravindakidambi  I tried to verify this change on local cluster and got below error in member cluster:
```
E1215 08:12:19.295339       1 remote_common_area.go:226] "error updating MemberClusterAnnounce" err="Internal error occurred: failed calling webhook \"mmemberclusterannounce.kb.io\": failed to call webhook: the server could not find the requested resource" Cluster=test-cluster-leader
```
I suppose you need a mutating webhook for memberannounce?

btw, please add sign-off info in your commits. thanks.

luolanzone commented:
@aravindakidambi  there are also a few github build error, please kindly fix them. thanks. https://github.com/antrea-io/antrea/runs/4528970206?check_suite_focus=true

luolanzone commented:
@aravindakidambi, I test this with resource pipeline, it looks fine, so I am OK to merge this. but one thing worth to be added as a TODO:
It's not related with this PR,  could it be possible to use back off and retry when remote common area is trying to `doMemberAnnounce()` in `Start()` but failed with error? for now it will just simply retry every 5s.

aravindakidambi commented:
> 

I fixed the build errors and also synced up with Lan, she tested again with new manifest from this branch and it worked fine. Previously manifest had a mutation webhook defined, but I removed it in this patch because there is no usecase to modify the request before persisting.
